drv/sidecar-mainboard-controller/src/tofino2.rs

```diff
@@ -231,3 +232,435 @@ impl Sequencer {
         })
     }
 }
+
+bitfield! {
+    #[derive(Copy, Clone, PartialEq, FromPrimitive, AsBytes)]
+    #[repr(C)]
+    pub struct DebugPortState(u8);
+    pub send_buffer_empty, set_send_buffer_empty: 0;
+    pub send_buffer_full, _: 1;
+    pub receive_buffer_empty, set_receive_buffer_empty: 2;
+    pub receive_buffer_full, _: 3;
+    pub request_in_progress, set_request_in_progress: 4;
+    pub address_nack_error, set_address_nack_error: 5;
+    pub byte_nack_error, set_byte_nack_error: 6;
+}
+
+#[derive(Copy, Clone, PartialEq, FromPrimitive, AsBytes)]
+#[repr(u8)]
+pub enum DebugRequestOpcode {
+    LocalWrite = 0,
+    LocalRead = 32,
+    DirectWrite = 128,
+    DirectRead = 160,
+    IndirectWrite = 192,
+    IndirectRead = 224,
+}
+
+impl From<DebugRequestOpcode> for u8 {
+    fn from(opcode: DebugRequestOpcode) -> Self {
+        opcode as u8
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, FromPrimitive, AsBytes)]
+#[repr(u8)]
+pub enum DirectBarSegment {
+    Bar0 = 0,
+    Msi = 1,
+    Cfg = 2,
+}
+
+impl From<DirectBarSegment> for u32 {
+    fn from(segment: DirectBarSegment) -> Self {
+        match segment {
+            DirectBarSegment::Bar0 => 0,
+            DirectBarSegment::Msi => (1 as u32) << 28,
+            DirectBarSegment::Cfg => (2 as u32) << 28,
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, FromPrimitive, AsBytes)]
+#[repr(u32)]
+pub enum TofinoRegisters {
+    Scratchpad = 0x0,
+    FreeRunningCounter = 0x10,
+    SpiOutData0 = (0x80000 | 0x120),
+    SpiOutData1 = (0x80000 | 0x124),
+    SpiInData = (0x80000 | 0x12c),
+    SpiCommand = (0x80000 | 0x128),
+    SpiIdCode = (0x80000 | 0x130),
+}
+
+impl From<TofinoRegisters> for u32 {
+    fn from(r: TofinoRegisters) -> Self {
+        r as u32
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, FromPrimitive, AsBytes)]
+#[repr(u8)]
+pub enum SpiEepromInstruction {
+    WriteEnable = 0x6,
+    WriteDisable = 0x4,
+    ReadStatusRegister = 0x5,
+    WriteStatusRegister = 0x1,
+    Read = 0x3,
+    Write = 0x2,
+}
+
+impl From<SpiEepromInstruction> for u8 {
+    fn from(i: SpiEepromInstruction) -> Self {
+        i as u8
+    }
+}
+
+impl From<SpiEepromInstruction> for u32 {
+    fn from(i: SpiEepromInstruction) -> Self {
+        i as u32
+    }
+}
+
+pub struct DebugPort {
+    fpga: FpgaUserDesign,
+}
+
+impl DebugPort {
+    pub fn new(task_id: userlib::TaskId) -> Self {
+        Self {
+            fpga: FpgaUserDesign::new(
+                task_id,
+                MainboardController::DEVICE_INDEX,
+            ),
+        }
+    }
+
+    pub fn state(&self) -> Result<DebugPortState, FpgaError> {
+        self.fpga
+            .read::<u8>(Addr::TOFINO_DEBUG_PORT_STATE)
+            .map(DebugPortState)
```

Comment:
If you add `FromBytes` to the `#[derive(...)]` block of `DebugPortState`, this could become simply
```rust
        self.fpga.read(Addr::TOFINO_DEBUG_PORT_STATE)
```
(since it would derive the return type from the function signature)